extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java

```diff
@@ -271,6 +285,101 @@ void registerMetricsFromAnnotatedMethods(SmallRyeMetricsRecorder metrics,
         }
     }
 
+    /**
+     * Mark metric producer methods and fields as unremovable, they should be kept even if
+     * there is no injection point for them.
+     */
+    @BuildStep
+    void unremovableProducers(BuildProducer<UnremovableBeanBuildItem> unremovable) {
+        Type type = Type.create(METRIC_INTERFACE, Type.Kind.CLASS);
+        unremovable.produce(
+                new UnremovableBeanBuildItem(new Predicate<io.quarkus.arc.processor.BeanInfo>() {
+                    @Override
+                    public boolean test(io.quarkus.arc.processor.BeanInfo beanInfo) {
+                        io.quarkus.arc.processor.BeanInfo declaringBean = beanInfo.getDeclaringBean();
+                        return (beanInfo.isProducerMethod() || beanInfo.isProducerField())
+                                && beanInfo.getTypes().contains(type)
+                                && declaringBean != null
+                                && declaringBean.getBeanClass() != null
+                                && !declaringBean.getBeanClass().toString().startsWith("io.smallrye.metrics");
+                    }
+                }));
+    }
+
+    @BuildStep
+    @Record(RUNTIME_INIT)
+    void registerMetricsFromProducers(
+            SmallRyeMetricsRecorder recorder,
+            ValidationPhaseBuildItem validationPhase,
+            BeanArchiveIndexBuildItem beanArchiveIndex) {
+        IndexView index = beanArchiveIndex.getIndex();
+        for (io.quarkus.arc.processor.BeanInfo bean : validationPhase.getContext().get(BuildExtension.Key.BEANS)) {
+            if (bean.isProducerField() || bean.isProducerMethod()) {
+                MetricType metricType = getMetricType(bean.getImplClazz());
+                if (metricType != null) {
+                    AnnotationTarget target = bean.getTarget().orElse(null);
```

Comment:
` bean.getTarget()` only returns null for synthetic beans.

Reply:
We should probably add some javadoc to `BeanInfo` methods ;-)

Reply:
Ok, just wanted to be extra safe, but if you're sure this will not be null then I'll remove the null checks :)